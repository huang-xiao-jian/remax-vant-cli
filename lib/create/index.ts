/**
 * @description - remant create <package> --ignore-page
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import path from 'path';
import fs from 'fs-extra';
import chalk from 'chalk';
import symbols from 'log-symbols';
import { format } from 'prettier';
import { pascalCase } from 'pascal-case';
import { compile } from 'handlebars';

export enum RenderType {
  INDEX = 'INDEX',
  PAGE = 'PAGE',
  COMPONENT = 'COMPONENT',
}

export interface CreateCommandOptions {
  ignorePage?: boolean;
  withoutDefaultProps?: boolean;
}

type RenderAbstract = (
  type: RenderType,
  name: string,
  options?: CreateCommandOptions
) => string;

export const render: RenderAbstract = (type, name, options) => {
  const location = path.resolve(__dirname, `./template/${type}.hbs`);
  const template = fs.readFileSync(location, 'utf-8');
  const context = options ? { ...options, name } : { name };
  const result = compile(template)(context);

  return format(result, {
    singleQuote: true,
    arrowParens: 'always',
    trailingComma: 'es5',
    parser: 'babel-ts',
  });
};

// eslint-disable-next-line import/prefer-default-export
export function create(component: string, options: CreateCommandOptions): void {
  const directory = process.cwd();

  const name = pascalCase(component);
  const files = [
    {
      filename: path.resolve(directory, `packages/${name}/index.ts`),
      content: render(RenderType.INDEX, name),
    },
    {
      filename: path.resolve(directory, `packages/${name}/${name}.tsx`),
      content: render(RenderType.COMPONENT, name),
    },
    {
      filename: path.resolve(directory, `packages/${name}/${name}.css`),
      content: '',
    },
  ];

  // several component without page preview
  if (!options.ignorePage) {
    files.push({
      filename: path.resolve(directory, `public/pages/${component}/index.tsx`),
      content: render(RenderType.PAGE, name, options),
    });
  }

  files.forEach((item) => {
    fs.ensureFile(item.filename).then(() => {
      // eslint-disable-next-line no-console
      console.log(
        ' ',
        symbols.success,
        chalk.cyan(path.relative(process.cwd(), item.filename))
      );

      fs.writeFileSync(item.filename, item.content);
    });
  });
}
