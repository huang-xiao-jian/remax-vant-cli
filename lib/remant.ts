#!/usr/bin/env node

import path from 'path';
import fs from 'fs-extra';
import chalk from 'chalk';
import symbols from 'log-symbols';
import { pascalCase } from 'pascal-case';
import { program } from 'commander';

// render split
const renderIndex = (name: string) => `
  /**
   * @description - nothing but export component
   */
  export { default } from './${name}';
`;
const renderComponent = (name: string) => `
// packages
import React, { FunctionComponent } from 'react';
import clsx from 'clsx';
import { View } from 'remax/wechat';
// internal
import './${name}.css';

// 默认值填充属性
interface Neutral${name}Props {}

interface Exogenous${name}Props {
  // 容器类名，用以覆盖内部
  className?: string;
}

type ${name}Props = Neutral${name}Props & Exogenous${name}Props;

const ${name}: FunctionComponent<${name}Props> = (props) => {
  const { className, children } = props;
  const classnames = {
    container: clsx(className),
  };

  return <View className={classnames.container}>{children}</View>;
};

export default ${name};
`;

const renderPage = (name: string) => `
// packages
import * as React from 'react';
import { View, Text } from 'remax/wechat';

// internal
import ${name} from '../../../packages/${name}';

export default () => {
  return (
    <View className="demo-block">
      <Text className="demo-block__title">基础用法</Text>
      <View className="demo-block__content">
      </View>
    </View>
  );
};
`;

const directory = process.cwd();

program
  .version('0.1.0')
  .command('create <component>')
  .option('-i, --ignore-page', 'create component without page')
  .description('setup package development grass')
  .action((component: string, option: { ignorePage: boolean }) => {
    const name = pascalCase(component);
    const files = [
      {
        filename: path.resolve(directory, `packages/${name}/index.ts`),
        content: renderIndex(name),
      },
      {
        filename: path.resolve(directory, `packages/${name}/${name}.tsx`),
        content: renderComponent(name),
      },
      {
        filename: path.resolve(directory, `packages/${name}/${name}.css`),
        content: '',
      },
    ];

    // several component without page preview
    if (!option.ignorePage) {
      files.push({
        filename: path.resolve(
          directory,
          `public/pages/${component}/index.tsx`
        ),
        content: renderPage(name),
      });
    }

    files.forEach((item) => {
      fs.ensureFile(item.filename).then(() => {
        console.log(
          ' ',
          symbols.success,
          chalk.cyan(path.relative(process.cwd(), item.filename))
        );

        fs.writeFileSync(item.filename, item.content);
      });
    });
  });

program.parse(process.argv);
