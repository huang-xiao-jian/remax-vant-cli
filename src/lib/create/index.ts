/**
 * @description - remant create <package> --ignore-page
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

// packages
import path from 'path';
import fs from 'fs-extra';
import chalk from 'chalk';
import { get } from 'node-emoji';
import { pascalCase } from 'pascal-case';
import { compile } from 'handlebars';

export enum RenderType {
  INDEX = 'index',
  PAGE = 'page',
  COMPONENT = 'component',
}

export interface CreateCommandOptions {
  ignorePage?: boolean;
  withoutDefaultProps?: boolean;
}

type RenderAbstract = (
  type: RenderType,
  name: string,
  options?: CreateCommandOptions
) => string;

export const render: RenderAbstract = (type, name, options) => {
  const location = path.resolve(__dirname, `./template/${type}.hbs`);
  const template = fs.readFileSync(location, 'utf-8');
  const context = options ? { ...options, name } : { name };

  return compile(template)(context);
};

// eslint-disable-next-line import/prefer-default-export
export function create(component: string, options: CreateCommandOptions): void {
  const project = process.cwd();

  const name = pascalCase(component);
  const files = [
    {
      filename: path.resolve(project, `packages/${name}/index.ts`),
      content: render(RenderType.INDEX, name),
    },
    {
      filename: path.resolve(project, `packages/${name}/${name}.tsx`),
      content: render(RenderType.COMPONENT, name),
    },
    {
      filename: path.resolve(project, `packages/${name}/${name}.css`),
      content: '',
    },
  ];

  // several component without page preview
  if (!options.ignorePage) {
    files.push({
      filename: path.resolve(project, `storyboard/pages/${component}/index.tsx`),
      content: render(RenderType.PAGE, name, options),
    });
  }

  files.forEach((item) => {
    fs.ensureFileSync(item.filename);

    // eslint-disable-next-line no-console
    console.log(' ', get('tada'), chalk.cyan(path.relative(project, item.filename)));

    fs.writeFileSync(item.filename, item.content);
  });
}
