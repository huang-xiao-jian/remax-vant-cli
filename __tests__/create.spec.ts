// internal
import { render, RenderType } from '../src/lib/create';

describe('create new component', () => {
  it('should render default export', () => {
    expect(render(RenderType.INDEX, 'Rate')).toMatchSnapshot();
  });

  it('should render page template', () => {
    expect(render(RenderType.PAGE, 'Rate')).toMatchSnapshot();
  });

  it('should render component template with default props', () => {
    expect(render(RenderType.COMPONENT, 'Rate')).toMatchSnapshot();
  });

  it('should render component template without default props', () => {
    expect(
      render(RenderType.COMPONENT, 'Rate', { withoutDefaultProps: true })
    ).toMatchSnapshot();
  });
});
