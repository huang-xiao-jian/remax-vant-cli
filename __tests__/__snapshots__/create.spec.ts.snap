// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create new component should render component template with default props 1`] = `
"// packages
import React, { FunctionComponent } from 'react';
import clsx from 'clsx';
import { View } from 'remax/wechat';
// internal
import withDefaultProps from '../tools/with-default-props-advance';
import './Rate.css';

// 默认值填充属性
interface NeutralRateProps {}

interface ExogenousRateProps {
  // 容器类名，用以覆盖内部
  className?: string;
}

type RateProps = NeutralRateProps & ExogenousRateProps;

// scope
const DefaultRateProps: NeutralRateProps = {};

const Rate: FunctionComponent<RateProps> = (props) => {
  const { className, children } = props;
  const classnames = {
    container: clsx(className),
  };

  return <View className={classnames.container}>{children}</View>;
};

export default withDefaultProps<ExogenousRateProps, NeutralRateProps>(DefaultRateProps)(Rate);
"
`;

exports[`create new component should render component template without default props 1`] = `
"// packages
import React, { FunctionComponent } from 'react';
import clsx from 'clsx';
import { View } from 'remax/wechat';
// internal
import './Rate.css';

interface RateProps {
  // 容器类名，用以覆盖内部
  className?: string;
}

const Rate: FunctionComponent<RateProps> = (props) => {
  const { className, children } = props;
  const classnames = {
    container: clsx(className),
  };

  return <View className={classnames.container}>{children}</View>;
};

export default Rate;
"
`;

exports[`create new component should render default export 1`] = `
"  /**
   * @description - nothing but export component
   */
  export { default } from './Rate';
"
`;

exports[`create new component should render page template 1`] = `
"// packages
import React from 'react';
import { View, Text } from 'remax/wechat';

// internal
import Rate from '../../../packages/Rate';

export default () => {
  return (
    <View className=\\"demo-block\\">
      <Text className=\\"demo-block__title\\">基础用法</Text>
      <View className=\\"demo-block__content\\">
      </View>
    </View>
  );
};
"
`;
